### PR TITLE
MS-Level and Precursor Information Written in TheoreticalSpectrumGenerator

### PR DESCRIPTION
**Pull request by Waschi97**
- Added MS-level and precursor information to `TSG::getSpectrum()`
- Added some tests for this
- Removed manual addition of those informations from `RawTandemMSSignalSimulation` and `TOPPViewBase`

### CODE REVIEW

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 1]
```diff
@@ -216,6 +216,21 @@ namespace OpenMS
...
 
     if (sort_by_position_) spectrum.sortByPositionPresorted(chunks.getChunks());
+
+    // set MS Level
+    spectrum.setMSLevel(2);
+
+    // set precursor
+    Precursor prec;
+
+    if (precursor_charge == 0)
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:226]: annotate that behaviour in doxygen

File: src/openms/source/SIMULATION/RawTandemMSSignalSimulation.cpp [hunk 2]
```diff
@@ -180,11 +180,11 @@ namespace OpenMS
         precomputed_MS2[i_f].push_back(p);
       }
 
-      precomputed_MS2[i_f].setMSLevel(2);
+      /*precomputed_MS2[i_f].setMSLevel(2);
```
Comment (cbielow) [inline src/openms/source/SIMULATION/RawTandemMSSignalSimulation.cpp:183]: omitting this does not work: the simulated spectrum is in `tmp_spec` and only the peaks are copied.
If you want to optimize, this will probably do it:
```
      Int prec_charge = features[i_f].getCharge();

      if (tandem_mode && svm_model_charges.count(prec_charge))
      {
        svm_spec_gen_set.simulate(precomputed_MS2[i_f], seq, rnd_gen_->getBiologicalRng(), prec_charge);
        precomputed_MS2[i_f].setMSLevel(2);
        Precursor prec;
        prec.setMZ(features[i_f].getMZ());
        precomputed_MS2[i_f].setPrecursors(std::vector<Precursor>(1, prec));
      }
      else
      {
        simple_generator.getSpectrum(precomputed_MS2[i_f], seq, 1, prec_charge, prec_charge);
      }

      precomputed_MS2[i_f].setMetaValue("FeatureID", static_cast<String>(features[i_f].getUniqueId()));
```
But you might as well undo the changes.

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 3]
```diff
@@ -2009,13 +2009,13 @@ namespace OpenMS
       }
 
       // set precursor information
-      vector<Precursor> precursors;
+      /*vector<Precursor> precursors;
```
Comment (cbielow) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:2012]: delete what is currently commented. It's just distracting, including `// set precursor information`.

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp [hunk 4]
```diff
@@ -129,10 +129,14 @@ START_SECTION(void getSpectrum(PeakSpectrum& spec, const AASequence& peptide, In
     TEST_REAL_SIMILAR(spec[i].getPosition()[0], result[i])
   }
 
+  TEST_REAL_SIMILAR(peptide.getMZ(2, Residue::Full), spec.getPrecursors()[0].getMZ());
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:132]: test the MS level as well?

File: src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp [hunk 5]
```diff
@@ -340,6 +346,8 @@ START_SECTION(void getSpectrum(PeakSpectrum& spec, const AASequence& peptide, In
   ptr->getSpectrum(spec, peptide, 3, 3);
   TEST_EQUAL(spec.size(), 30)
 
+  TEST_REAL_SIMILAR(peptide.getMZ(4, Residue::Full), spec.getPrecursors()[0].getMZ());
+
```
Comment (cbielow) [inline src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:350]: do repetitive calls to `.getSpectrum()` and check if you get additional PC's

Comment (cbielow) [reply src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:350]: and peaks.

**Comment by timosachsenberg** [comment]
does it make sense to add some assertions like max_charge can't be larger than precursor_charge (except if pc == 0)

**Comment by cbielow** [comment]
ProteomicsLFQ test is failing
https://abibuilder.informatik.uni-tuebingen.de/jenkins/job/openms/job/PR/job/bldtst/compiler=g++-7,os_label=ubuntu_1804/755/testReport/projectroot.src.tests/topp/UTILS_ProteomicsLFQ_3/

@jpfeuffer did you have a fix for that already?

**Comment by jpfeuffer** [comment]
Yes it is in the huge PR, you just need the  use_ion_series_score parameter from that (and set it to false in FFID) as a quick fix. The PR fixes the problem and related ones a bit more elaborately.

**Comment by cbielow** [comment]
compile error: http://cdash.openms.de/viewBuildError.php?buildid=96716

**Comment by cbielow** [comment]
rebuild jenkins

**Comment by jpfeuffer** [comment]
rebuild jenkins

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 6]
```diff
@@ -216,6 +216,29 @@ namespace OpenMS
...
+    {
+      precursor_charge = max_charge +1;
+    }
+    
+    if (precursor_charge <= max_charge)
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:234]: this is not correct, IMHO.
In MALDI, all charges are 1.

So, it should be `if (precursor_charge < max_charge)` 
Please also add a 
`@throw Exception::InvalidParameter if precursor_charge < max_charge

Comment (timosachsenberg) [reply src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:234]: good point

**Review by jpfeuffer — APPROVED** [review]

**Comment by timosachsenberg** [comment]
we should not forget to reenable the test in the fix
